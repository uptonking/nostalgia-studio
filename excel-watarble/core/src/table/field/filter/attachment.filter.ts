import { z } from 'zod';
import { baseFilter } from './filter.base';
import { attachmentFilterOperators } from './operators';

const attachmentFilterTypeValue = z.enum([
  'image',
  'video',
  'text',
  'document',
  'excel',
  'ppt',
  'pdf',
]);
export type IAttachmentFilterTypeValue = z.infer<
  typeof attachmentFilterTypeValue
>;
export const attachmentFilterValue = z
  .string()
  .nullable()
  .or(attachmentFilterTypeValue);

export const attachmentFilter = z
  .object({
    type: z.literal('attachment'),
    operator: attachmentFilterOperators,
    value: attachmentFilterValue,
  })
  .merge(baseFilter);

export type IAttachmentFilter = z.infer<typeof attachmentFilter>;
export type IAttachmentFilterOperator = z.infer<
  typeof attachmentFilterOperators
>;
