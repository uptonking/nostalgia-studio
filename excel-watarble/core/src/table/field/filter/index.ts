export * from './filter';
export * from './operators';
export * from './root-filter';
