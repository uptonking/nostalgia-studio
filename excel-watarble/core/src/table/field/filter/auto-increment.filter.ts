import { z } from 'zod';
import { baseFilter } from './filter.base';
import { autoIncrementFilterOperators } from './operators';

export const autoIncrementFilterValue = z.number().nullable();
export const autoIncrementFilter = z
  .object({
    type: z.literal('auto-increment'),
    operator: autoIncrementFilterOperators,
    value: autoIncrementFilterValue,
  })
  .merge(baseFilter);

export type IAutoIncrementFilter = z.infer<typeof autoIncrementFilter>;
