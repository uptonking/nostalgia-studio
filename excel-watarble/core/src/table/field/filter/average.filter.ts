import { z } from 'zod';
import { baseFilter } from './filter.base';
import { averageFilterOperators } from './operators';

export const averageFilterValue = z.number().nullable();
export const averageFilter = z
  .object({
    type: z.literal('average'),
    operator: averageFilterOperators,
    value: averageFilterValue,
  })
  .merge(baseFilter);
export type IAverageFilter = z.infer<typeof averageFilter>;
export type IAverageFilterOperator = z.infer<typeof averageFilterOperators>;
