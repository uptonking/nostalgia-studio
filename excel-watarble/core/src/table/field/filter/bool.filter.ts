import { z } from 'zod';
import { baseFilter } from './filter.base';
import { boolFilterOperators } from './operators';

export const boolFilterValue = z.boolean().nullable();
export const boolFilter = z
  .object({
    type: z.literal('bool'),
    operator: boolFilterOperators,
    value: boolFilterValue,
  })
  .merge(baseFilter);

export type IBoolFilter = z.infer<typeof boolFilter>;
