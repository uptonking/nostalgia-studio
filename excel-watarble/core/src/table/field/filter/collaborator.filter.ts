import { z } from 'zod';

import { collaboratorFieldValue } from '../collaborator-field.type';
import { baseFilter } from './filter.base';
import { collaboratorFilterOperators } from './operators';

export const collaboratorFilterValue = collaboratorFieldValue;
export const collaboratorFilter = z
  .object({
    type: z.literal('collaborator'),
    operator: collaboratorFilterOperators,
    value: collaboratorFilterValue,
  })
  .merge(baseFilter);

export type ICollaboratorFilter = z.infer<typeof collaboratorFilter>;
export type ICollaboratorFilterOperator = z.infer<
  typeof collaboratorFilterOperators
>;
