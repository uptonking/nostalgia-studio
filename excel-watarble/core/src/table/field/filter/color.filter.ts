import { z } from 'zod';

import { colorFieldValue } from '../color-field.type';
import { baseFilter } from './filter.base';
import { colorFilterOperators } from './operators';

export const colorFilterValue = colorFieldValue;
export const colorFilter = z
  .object({
    type: z.literal('color'),
    operator: colorFilterOperators,
    value: colorFilterValue,
  })
  .merge(baseFilter);

export type IColorFilter = z.infer<typeof colorFilter>;
export type IColorFilterOperator = z.infer<typeof colorFilterOperators>;
