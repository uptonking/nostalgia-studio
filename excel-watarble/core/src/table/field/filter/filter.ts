import { type Option } from 'oxide.ts';
import { None, Some } from 'oxide.ts';
import { z } from 'zod';

import { type CompositeSpecification } from '@datalking/pivot-entity';

import {
  BoolIsFalse,
  BoolIsTrue,
  DateEqual,
  DateGreaterThan,
  DateGreaterThanOrEqual,
  DateIsToday,
  DateLessThan,
  DateLessThanOrEqual,
  DateRangeEqual,
  HasExtension,
  HasFileType,
  IsAttachmentEmpty,
  IsTreeRoot,
  NumberEqual,
  NumberGreaterThan,
  NumberGreaterThanOrEqual,
  NumberLessThan,
  NumberLessThanOrEqual,
  SelectEqual,
  SelectIn,
  StringContain,
  StringEndsWith,
  StringEqual,
  StringRegex,
  StringStartsWith,
  WithRecordIds,
} from '../../record/index';
import { colorFieldValue } from '../color-field.type';
import {
  DateFieldValue,
  NumberFieldValue,
  SelectFieldValue,
  StringFieldValue,
} from '../index';
import { type ISelectFieldValue } from '../select-field.type';
import {
  type IAttachmentFilter,
  type IAttachmentFilterTypeValue,
} from './attachment.filter';
import { attachmentFilter, attachmentFilterValue } from './attachment.filter';
import { type IAutoIncrementFilter } from './auto-increment.filter';
import {
  autoIncrementFilter,
  autoIncrementFilterValue,
} from './auto-increment.filter';
import { type IAverageFilter } from './average.filter';
import { averageFilter, averageFilterValue } from './average.filter';
import { type IBoolFilter } from './bool.filter';
import { boolFilter, boolFilterValue } from './bool.filter';
import {
  collaboratorFilter,
  collaboratorFilterValue,
} from './collaborator.filter';
import { type IColorFilter } from './color.filter';
import { colorFilter } from './color.filter';
import { type IConjunction } from './conjunction';
import { conjunctions } from './conjunction';
import { type ICountFilter } from './count.filter';
import { countFilter, countFilterValue } from './count.filter';
import { type ICreatedAtFilter } from './created-at.filter';
import { createdAtFilter, createdAtFilterValue } from './created-at.filter';
import { createdByFilter, createdByFilterValue } from './created-by.filter';
import { type IDateRangeFilter } from './date-range.filter';
import { dateRangeFilter, dateRangeFilterValue } from './date-range.filter';
import { type IDateFilter } from './date.filter';
import { dateFilter, dateFilterValue } from './date.filter';
import { type IEmailFilter } from './email.filter';
import { emailFilter, emailFilterValue } from './email.filter';
import { type IIdFilter } from './id.filter';
import { idFilter, idFilterValue } from './id.filter';
import { lookupFilter, lookupFilterValue } from './lookup.filter';
import { type INumberFilter } from './number.filter';
import { numberFilter, numberFilterValue } from './number.filter';
import {
  $eq,
  $is_false,
  $is_root,
  $is_today,
  $is_true,
  $neq,
  attachmentFilterOperators,
  autoIncrementFilterOperators,
  averageFilterOperators,
  boolFilterOperators,
  collaboratorFilterOperators,
  colorFilterOperators,
  countFilterOperators,
  createdAtFilterOperators,
  createdByFilterOperators,
  dateFilterOperators,
  dateRangeFilterOperators,
  emailFilterOperators,
  idFilterOperators,
  lookupFilterOperators,
  numberFilterOperators,
  parentFilterOperators,
  ratingFilterOperators,
  referenceFilterOperators,
  selectFilterOperators,
  stringFilterOperators,
  sumFilterOperators,
  treeFilterOperators,
  updatedAtFilterOperators,
  updatedByFilterOperators,
} from './operators';
import { parentFilter, parentFilterValue } from './parent.filter';
import { type IRatingFilter } from './rating.filter';
import { ratingFilter, ratingFilterValue } from './rating.filter';
import { referenceFilter, referenceFilterValue } from './reference.filter';
import { type ISelectFilter } from './select.filter';
import { selectFilter, selectFilterValue } from './select.filter';
import { type IStringFilter } from './string.filter';
import { stringFilter, stringFilterValue } from './string.filter';
import { type ISumFilter } from './sum.filter';
import { sumFilter, sumFilterValue } from './sum.filter';
import { type ITreeFilter } from './tree.filter';
import { treeFilter, treeFilterValue } from './tree.filter';
import { type IUpdatedAtFilter } from './updated-at.filter';
import { updatedAtFilter, updatedAtFilterValue } from './updated-at.filter';
import { updatedByFilter, updatedByFilterValue } from './updated-by.filter';

export const filterValue = z.union([
  idFilterValue,
  createdAtFilterValue,
  updatedAtFilterValue,
  autoIncrementFilterValue,
  stringFilterValue,
  emailFilterValue,
  colorFieldValue,
  numberFilterValue,
  dateFilterValue,
  dateRangeFilterValue,
  selectFilterValue,
  boolFilterValue,
  referenceFilterValue,
  treeFilterValue,
  parentFilterValue,
  ratingFilterValue,
  countFilterValue,
  lookupFilterValue,
  sumFilterValue,
  averageFilterValue,
  attachmentFilterValue,
  collaboratorFilterValue,
  createdByFilterValue,
  updatedByFilterValue,
]);
export type IFilterValue = z.infer<typeof filterValue>;

export const operaotrs = z.union([
  idFilterOperators,
  createdAtFilterOperators,
  updatedAtFilterOperators,
  autoIncrementFilterOperators,
  stringFilterOperators,
  emailFilterOperators,
  colorFilterOperators,
  numberFilterOperators,
  dateFilterOperators,
  dateRangeFilterOperators,
  selectFilterOperators,
  boolFilterOperators,
  referenceFilterOperators,
  treeFilterOperators,
  parentFilterOperators,
  ratingFilterOperators,
  countFilterOperators,
  lookupFilterOperators,
  sumFilterOperators,
  averageFilterOperators,
  attachmentFilterOperators,
  collaboratorFilterOperators,
  createdByFilterOperators,
  updatedByFilterOperators,
]);
export type IOperator = z.infer<typeof operaotrs>;

const filter = z.discriminatedUnion('type', [
  idFilter,
  createdAtFilter,
  updatedAtFilter,
  autoIncrementFilter,
  stringFilter,
  emailFilter,
  colorFilter,
  numberFilter,
  dateFilter,
  dateRangeFilter,
  selectFilter,
  boolFilter,
  referenceFilter,
  treeFilter,
  parentFilter,
  ratingFilter,
  countFilter,
  lookupFilter,
  sumFilter,
  averageFilter,
  attachmentFilter,
  collaboratorFilter,
  createdByFilter,
  updatedByFilter,
]);

export type IFilter = z.infer<typeof filter>;
export type IFilters = IFilter[];

export interface IGroup {
  conjunction: IConjunction;
  children?: IFilterOrGroupList;
}

const group = z.lazy(() =>
  z.object({
    conjunction: conjunctions,
    children: z.union([group, filter]).array().nonempty().optional(),
  }),
) as z.ZodType<IGroup>;

const filterOrGroup = filter.or(group);
export type IFilterOrGroup = z.infer<typeof filterOrGroup>;

export const filterOrGroupList = filterOrGroup.array();
export type IFilterOrGroupList = z.infer<typeof filterOrGroupList>;
export const rootFilter = filterOrGroup.or(filterOrGroupList);
export type IRootFilter = z.infer<typeof rootFilter>;

const isGroup = (filterOrGroup: IFilterOrGroup): filterOrGroup is IGroup => {
  return Object.hasOwn(filterOrGroup, 'conjunction');
};

const isFilter = (filterOrGroup: IFilterOrGroup): filterOrGroup is IFilter => {
  return (
    Object.hasOwn(filterOrGroup, 'type') &&
    Object.hasOwn(filterOrGroup, 'operator')
  );
};

const convertIdFilter = (filter: IIdFilter): Option<CompositeSpecification> => {
  if (filter.value === undefined) {
    return None;
  }

  switch (filter.operator) {
    case '$eq': {
      return Some(
        new StringEqual(
          filter.path,
          new StringFieldValue(filter.value as string),
        ),
      );
    }
    case '$neq': {
      return Some(
        new StringEqual(
          filter.path,
          new StringFieldValue(filter.value as string),
        ).not(),
      );
    }
    case '$in': {
      return Some(WithRecordIds.fromIds(filter.value as string[]));
    }
    case '$nin': {
      return Some(WithRecordIds.fromIds(filter.value as string[]).not());
    }

    default:
      return None;
  }
};

const convertStringFilter = (
  filter: IStringFilter | IEmailFilter | IColorFilter,
): Option<CompositeSpecification> => {
  if (filter.value === undefined) {
    return None;
  }

  switch (filter.operator) {
    case '$eq': {
      return Some(
        new StringEqual(filter.path, new StringFieldValue(filter.value)),
      );
    }
    case '$neq': {
      return Some(
        new StringEqual(filter.path, new StringFieldValue(filter.value)).not(),
      );
    }
    case '$contains': {
      return Some(
        new StringContain(filter.path, new StringFieldValue(filter.value)),
      );
    }
    case '$starts_with': {
      return Some(
        new StringStartsWith(filter.path, new StringFieldValue(filter.value)),
      );
    }
    case '$ends_with': {
      return Some(
        new StringEndsWith(filter.path, new StringFieldValue(filter.value)),
      );
    }
    case '$regex': {
      return Some(
        new StringRegex(filter.path, new StringFieldValue(filter.value)),
      );
    }

    default:
      return None;
  }
};

const convertNumberFilter = (
  filter:
    | INumberFilter
    | IAutoIncrementFilter
    | IRatingFilter
    | ICountFilter
    | ISumFilter
    | IAverageFilter,
): Option<CompositeSpecification> => {
  if (filter.value === undefined) {
    return None;
  }

  switch (filter.operator) {
    case '$eq': {
      return Some(
        new NumberEqual(filter.path, new NumberFieldValue(filter.value)),
      );
    }
    case '$neq': {
      return Some(
        new NumberEqual(filter.path, new NumberFieldValue(filter.value)).not(),
      );
    }
    case '$gt': {
      return Some(
        new NumberGreaterThan(filter.path, new NumberFieldValue(filter.value)),
      );
    }
    case '$gte': {
      return Some(
        new NumberGreaterThanOrEqual(
          filter.path,
          new NumberFieldValue(filter.value),
        ),
      );
    }
    case '$lt': {
      return Some(
        new NumberLessThan(filter.path, new NumberFieldValue(filter.value)),
      );
    }
    case '$lte': {
      return Some(
        new NumberLessThanOrEqual(
          filter.path,
          new NumberFieldValue(filter.value),
        ),
      );
    }
    default:
      return None;
  }
};

const convertSelectFilter = (
  filter: ISelectFilter,
): Option<CompositeSpecification> => {
  if (filter.value === undefined) {
    return None;
  }

  switch (filter.operator) {
    case '$eq': {
      return Some(
        new SelectEqual(
          filter.path,
          new SelectFieldValue(filter.value as ISelectFieldValue),
        ),
      );
    }
    case '$neq': {
      return Some(
        new SelectEqual(
          filter.path,
          new SelectFieldValue(filter.value as ISelectFieldValue),
        ).not(),
      );
    }
    case '$in': {
      return Some(
        new SelectIn(
          filter.path,
          (filter.value as ISelectFieldValue[]).map(
            (v) => new SelectFieldValue(v),
          ),
        ),
      );
    }
    case '$nin': {
      return Some(
        new SelectIn(
          filter.path,
          (filter.value as ISelectFieldValue[]).map(
            (v) => new SelectFieldValue(v),
          ),
        ).not(),
      );
    }

    default: {
      return None;
    }
  }
};

const convertBoolFilter = (
  filter: IBoolFilter,
): Option<CompositeSpecification> => {
  switch (filter.operator) {
    case $is_true.value: {
      return Some(new BoolIsTrue(filter.path));
    }
    case $is_false.value: {
      return Some(new BoolIsFalse(filter.path));
    }

    default: {
      return None;
    }
  }
};

const convertDateRangeFilter = (
  filter: IDateRangeFilter,
): Option<CompositeSpecification> => {
  switch (filter.operator) {
    case $eq.value:
      return Some(DateRangeEqual.fromString(filter.path, filter.value as any));
    case $neq.value:
      return Some(
        DateRangeEqual.fromString(filter.path, filter.value as any).not(),
      );

    default:
      return None;
  }
};

const convertDateFilter = (
  filter: IDateFilter | ICreatedAtFilter | IUpdatedAtFilter,
): Option<CompositeSpecification> => {
  if (filter.operator === $is_today.value) {
    return Some(new DateIsToday(filter.path));
  }

  if (filter.value === undefined) {
    return None;
  }

  switch (filter.operator) {
    case '$eq': {
      return Some(
        new DateEqual(
          filter.path,
          DateFieldValue.fromNullableString(filter.value as string),
        ),
      );
    }
    case '$neq': {
      return Some(
        new DateEqual(
          filter.path,
          DateFieldValue.fromNullableString(filter.value as string),
        ).not(),
      );
    }
    case '$gt': {
      return Some(
        new DateGreaterThan(
          filter.path,
          DateFieldValue.fromString(filter.value as string),
        ),
      );
    }
    case '$gte': {
      return Some(
        new DateGreaterThanOrEqual(
          filter.path,
          DateFieldValue.fromString(filter.value as string),
        ),
      );
    }
    case '$lt': {
      return Some(
        new DateLessThan(
          filter.path,
          DateFieldValue.fromString(filter.value as string),
        ),
      );
    }
    case '$lte': {
      return Some(
        new DateLessThanOrEqual(
          filter.path,
          DateFieldValue.fromString(filter.value as string),
        ),
      );
    }
    default:
      return None;
  }
};

const convertTreeFilter = (
  filter: ITreeFilter,
): Option<CompositeSpecification> => {
  switch (filter.operator) {
    case $is_root.value: {
      return Some(new IsTreeRoot(filter.path));
    }

    default: {
      return None;
    }
  }
};

const convertAttachmentFilter = (
  filter: IAttachmentFilter,
): Option<CompositeSpecification> => {
  switch (filter.operator) {
    case '$has_file_type':
      return Some(
        new HasFileType(
          filter.path,
          filter.value as IAttachmentFilterTypeValue,
        ),
      );
    case '$is_empty':
      return Some(new IsAttachmentEmpty(filter.path, undefined));
    case '$is_not_empty':
      return Some(new IsAttachmentEmpty(filter.path, undefined).not());
    case '$has_file_extension':
      return Some(new HasExtension(filter.path, filter.value as string));
  }
};

const convertFilter = (filter: IFilter): Option<CompositeSpecification> => {
  switch (filter.type) {
    case 'id':
      return convertIdFilter(filter);
    case 'string':
    case 'email':
    case 'color':
      return convertStringFilter(filter);
    case 'number':
    case 'rating':
    case 'auto-increment':
    case 'count':
    case 'sum':
    case 'average':
      return convertNumberFilter(filter);
    case 'date':
    case 'created-at':
    case 'updated-at':
      return convertDateFilter(filter);
    case 'date-range':
      return convertDateRangeFilter(filter);
    case 'select':
      return convertSelectFilter(filter);
    case 'bool':
      return convertBoolFilter(filter);
    case 'reference':
      throw new Error('convertFilter.reference not implemented');
    case 'tree':
      return convertTreeFilter(filter);
    case 'attachment':
      return convertAttachmentFilter(filter);
    default:
      return None;
  }
};

const convertFilterOrGroup = (
  filterOrGroup: IFilterOrGroup,
): Option<CompositeSpecification> => {
  if (isGroup(filterOrGroup)) {
    return convertFilterOrGroupList(
      filterOrGroup.children,
      filterOrGroup.conjunction,
    );
  } else if (isFilter(filterOrGroup)) {
    return convertFilter(filterOrGroup);
  }

  return None;
};

const convertFilterOrGroupList = (
  filterOrGroupList: IFilterOrGroupList = [],
  conjunction: IConjunction = '$and',
): Option<CompositeSpecification> => {
  let spec: Option<CompositeSpecification> = None;
  for (const filter of filterOrGroupList) {
    if (spec.isNone()) {
      spec = convertFilterOrGroup(filter);
      if (conjunction === '$not') {
        return spec.map((s) => s.not());
      }
    } else {
      if (isFilter(filter)) {
        spec = spec.map((left) => {
          const right = convertFilterOrGroup(filter);
          if (right.isSome()) {
            if (conjunction === '$and') {
              return left.and(right.unwrap());
            } else if (conjunction === '$or') {
              return left.or(right.unwrap());
            }
            return left.and(right.unwrap().not());
          }
          return left;
        });
      } else if (isGroup(filter)) {
        spec = convertFilterOrGroupList(filter.children, filter.conjunction);
      }
    }
  }

  return spec;
};

export const convertFilterSpec = (
  filter: IRootFilter,
): Option<CompositeSpecification> => {
  if (Array.isArray(filter)) {
    return convertFilterOrGroupList(filter);
  }

  return convertFilterOrGroup(filter);
};
