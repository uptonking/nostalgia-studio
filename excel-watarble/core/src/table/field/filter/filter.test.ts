import { convertFilterSpec, type IRootFilter, rootFilter } from './filter';

test.each<IRootFilter>([
  {
    type: 'string',
    operator: '$eq',
    path: 'name',
    value: 'hello',
  },
  {
    type: 'string',
    operator: '$contains',
    path: 'name',
    value: 'hello',
  },
  {
    type: 'string',
    operator: '$starts_with',
    path: 'name',
    value: 'starts with',
  },
  {
    type: 'string',
    operator: '$ends_with',
    path: 'name',
    value: 'ends with',
  },
  {
    type: 'number',
    operator: '$neq',
    path: 'name',
    value: 1,
  },
  {
    type: 'number',
    operator: '$neq',
    path: 'name.nested',
    value: 1,
  },
  [
    {
      conjunction: '$and',
      children: [
        {
          type: 'string',
          operator: '$eq',
          path: 'name',
          value: 'hello',
        },
      ],
    },
  ],
  [
    {
      conjunction: '$or',
      children: [
        {
          type: 'string',
          operator: '$eq',
          path: 'name',
          value: 'hello',
        },
      ],
    },
  ],
  {
    type: 'number',
    operator: '$gt',
    path: 'gt',
    value: 1,
  },
  {
    type: 'number',
    operator: '$gte',
    path: 'gte',
    value: 1,
  },
  {
    type: 'number',
    operator: '$lt',
    path: 'lt',
    value: 1,
  },
  {
    type: 'number',
    operator: '$lte',
    path: 'lte',
    value: 1,
  },
  {
    type: 'date',
    operator: '$is_today',
    path: 'is_today',
    value: null,
  },
  {
    type: 'bool',
    operator: '$is_true',
    path: 'field',
    value: true,
  },
  {
    type: 'bool',
    operator: '$is_false',
    path: 'field',
    value: false,
  },
])('should create root filter', (filter) => {
  const parsed = rootFilter.parse(filter);
  expect(parsed).toEqual(filter);

  const spec = convertFilterSpec(filter);
  expect(spec).toMatchSnapshot();
});
