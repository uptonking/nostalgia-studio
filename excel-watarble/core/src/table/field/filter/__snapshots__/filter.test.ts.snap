// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should create root filter 1`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEqual {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "hello",
      },
    },
  },
}
`;

exports[`should create root filter 2`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringContain {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "hello",
      },
    },
  },
}
`;

exports[`should create root filter 3`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringStartsWith {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "starts with",
      },
    },
  },
}
`;

exports[`should create root filter 4`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEndsWith {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "ends with",
      },
    },
  },
}
`;

exports[`should create root filter 5`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): T {
    "spec": NumberEqual {
      "fieldId": "name",
      "value": NumberFieldValue {
        "props": {
          "value": 1,
        },
      },
    },
  },
}
`;

exports[`should create root filter 6`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): T {
    "spec": NumberEqual {
      "fieldId": "name.nested",
      "value": NumberFieldValue {
        "props": {
          "value": 1,
        },
      },
    },
  },
}
`;

exports[`should create root filter 7`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEqual {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "hello",
      },
    },
  },
}
`;

exports[`should create root filter 8`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEqual {
    "fieldId": "name",
    "value": StringFieldValue {
      "props": {
        "value": "hello",
      },
    },
  },
}
`;

exports[`should create root filter 9`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberGreaterThan {
    "fieldId": "gt",
    "value": NumberFieldValue {
      "props": {
        "value": 1,
      },
    },
  },
}
`;

exports[`should create root filter 10`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberGreaterThanOrEqual {
    "fieldId": "gte",
    "value": NumberFieldValue {
      "props": {
        "value": 1,
      },
    },
  },
}
`;

exports[`should create root filter 11`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberLessThan {
    "fieldId": "lt",
    "value": NumberFieldValue {
      "props": {
        "value": 1,
      },
    },
  },
}
`;

exports[`should create root filter 12`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberLessThanOrEqual {
    "fieldId": "lte",
    "value": NumberFieldValue {
      "props": {
        "value": 1,
      },
    },
  },
}
`;

exports[`should create root filter 13`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateIsToday {
    "fieldId": "is_today",
    "value": DateFieldValue {
      "props": {
        "value": null,
      },
    },
  },
}
`;

exports[`should create root filter 14`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): BoolIsTrue {
    "fieldId": "field",
    "value": BoolFieldValue {
      "props": {
        "value": true,
      },
    },
  },
}
`;

exports[`should create root filter 15`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): BoolIsFalse {
    "fieldId": "field",
    "value": BoolFieldValue {
      "props": {
        "value": false,
      },
    },
  },
}
`;
